testing/sdk_upgrade.py

```diff
@@ -89,131 +98,158 @@ def test_upgrade(
-        update_cmd.append('--package-version={}'.format(to_package_version))
-    if additional_options:
-        options_file = tempfile.NamedTemporaryFile("w")
-        json.dump(additional_options, options_file)
-        options_file.flush()  # ensure json content is available for the CLI to read below
-        update_cmd.append("--options={}".format(options_file.name))
-    sdk_cmd.svc_cli(package_name, service_name, ' '.join(update_cmd), check=True)
-
-
-def _upgrade_or_downgrade(
-        package_name,
-        to_package_version,
-        service_name,
-        running_task_count,
-        additional_options,
-        timeout_seconds,
-        wait_for_deployment):
-
+def update_or_upgrade_or_downgrade(
+    package_name,
+    service_name,
+    to_package_version,
+    additional_options,
+    expected_running_tasks,
+    wait_for_deployment=True,
+    timeout_seconds=sdk_utils.TIMEOUT_SECONDS,
+):
     initial_config = get_config(package_name, service_name)
-    task_ids = sdk_tasks.get_task_ids(service_name, '')
-
-    if sdk_utils.dcos_version_less_than("1.10") or shakedown.ee_version() is None:
-        log.info('Using marathon upgrade flow to upgrade {} {}'.format(package_name, to_package_version))
+    task_ids = sdk_tasks.get_task_ids(service_name, "")
+    if (to_package_version and not sdk_utils.is_cli_supports_service_version_upgrade()) or (
+        additional_options and not sdk_utils.is_cli_supports_service_options_update()
+    ):
+        log.info(
+            "Using marathon flow to upgrade to package name : {} version : [{}]".format(
+                package_name, to_package_version
+            )
+        )
         sdk_marathon.destroy_app(service_name)
         sdk_install.install(
             package_name,
             service_name,
-            running_task_count,
+            expected_running_tasks,
             additional_options=additional_options,
             package_version=to_package_version,
             timeout_seconds=timeout_seconds,
-            wait_for_deployment=wait_for_deployment)
+            wait_for_deployment=wait_for_deployment,
+        )
     else:
-        log.info('Using CLI upgrade flow to upgrade {} {}'.format(package_name, to_package_version))
-        update_service(package_name, service_name, additional_options, to_package_version)
+        _update_service_with_cli(package_name, service_name, to_package_version, additional_options)
+    return not wait_for_deployment or _wait_for_deployment(
+        package_name, service_name, initial_config, task_ids, timeout_seconds
+    )
+
+
+def _update_service_with_cli(
+    package_name, service_name, to_package_version=None, additional_options=None
+):
+    log.info(
+        "Using CLI to update to package name : {} version : [{}]".format(
+            package_name, to_package_version
+        )
+    )
+    update_cmd = ["update", "start"]
+    if to_package_version:
+        sdk_utils.ensure_cli_supports_service_version_upgrade()
```

Comment:
Would failing this check effectively imply a bug in `sdk_upgrade`, or should the caller have done something differently? If it's an internal error I think it'd make sense to mention something like "INTERNAL ERROR: ..." in the assert message (once that assert is moved here from `sdk_utils` per comment there)

(Same question for L212)

Reply:
The caller should have done something differently. Since we have `is_cli_supports_*` methods that return boolean, the caller can use that to decide what to do "differently". The assertion error messages are to let the caller know what needs to be done differently.